cpuid/src/common.rs

```diff
@@ -0,0 +1,211 @@
+    let max_function = unsafe { __get_cpuid_max(function & EXT_FUNCTION).0 };
+    if function > max_function {
+        return Err(Error::InvalidParameters(format!(
+            "Function not supported: 0x{:x}",
+            function
+        )));
+    }
+
+    // this is safe because the host supports the `cpuid` instruction
+    let entry = unsafe { __cpuid_count(function, count) };
+    if entry.eax == 0 && entry.ebx == 0 && entry.ecx == 0 && entry.edx == 0 {
+        return Err(Error::InvalidParameters(format!(
+            "Invalid count: {}",
+            count
+        )));
+    }
+
+    Ok(entry)
+}
+
+/// Extracts the CPU vendor id from leaf 0x0.
+///
+#[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+pub fn get_vendor_id() -> Result<[u8; 12], Error> {
+    match get_cpuid(0, 0) {
+        Ok(vendor_entry) => {
+            let bytes: [u8; 12] = unsafe {
+                std::mem::transmute([vendor_entry.ebx, vendor_entry.edx, vendor_entry.ecx])
+            };
+            Ok(bytes)
+        }
+        Err(_e) => Err(Error::NotSupported),
+    }
+}
+
+/// Replaces the `cpuid` entries corresponding to `function` with the entries from the host's cpuid.
+///
+#[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+pub fn emulate_host_cpuid_function(cpuid: &mut CpuId, function: u32) -> Result<(), Error> {
+    // copy all the CpuId entries, except for the ones with the provided function
+    let mut entries: Vec<kvm_cpuid_entry2> = Vec::new();
+    for entry in cpuid.mut_entries_slice().iter() {
+        if entry.function != function {
+            entries.push(*entry);
+        }
+    }
+
+    // add all the host leaves with the provided function
+    let mut count: u32 = 0;
+    while let Ok(entry) = get_cpuid(function, count) {
+        // check if there's enough space to add a new entry to the cpuid
+        if entries.len() == kvm::MAX_KVM_CPUID_ENTRIES {
+            return Err(Error::SizeLimitExceeded);
+        }
+
+        entries.push(kvm_cpuid_entry2 {
+            function: function,
+            index: count,
+            flags: 0,
+            eax: entry.eax,
+            ebx: entry.ebx,
+            ecx: entry.ecx,
+            edx: entry.edx,
+            padding: [0, 0, 0],
+        });
+        count += 1;
+    }
+
+    let cpuid2 = CpuId::from_entries(&entries);
+    *cpuid = cpuid2;
+
+    //    cpuid.set_entries(&entries);
+    Ok(())
+}
+
+#[cfg(test)]
+mod tests {
+    use common::*;
+    use kvm::CpuId;
+
+    #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+    fn get_topoext_fn() -> u32 {
+        let vendor_id = get_vendor_id();
+        assert!(vendor_id.is_ok());
+        let function = match &vendor_id.ok().unwrap() {
+            INTEL => 0x4,
+            AMD => 0x8000001d,
+            _ => 0,
+        };
+        assert!(function != 0);
+
+        function
+    }
+
+    #[test]
+    #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+    fn get_cpu_id_test() {
```

Comment:
Please prefix test function with the word `test` so it is easier to mentally separate the tests from the helper functions.